File: examples/anip/anip_server/primitives/delegation.py
"""Delegation chain validation.

Validates the full delegation DAG: token expiry, scope sufficiency,
purpose binding, delegation depth, and concurrent branch constraints.

Tokens are persisted in SQLite for durability and audit trail.
"""

from __future__ import annotations

from datetime import datetime, timezone

from ..data.database import load_token as db_load_token
from ..data.database import store_token as db_store_token
from .models import ANIPFailure, ConcurrentBranches, DelegationToken, Resolution

# Active request tracking for concurrent_branches enforcement
_active_requests: set[str] = set()


def register_token(token: DelegationToken) -> None:
    """Register a delegation token, persisting to SQLite."""
    db_store_token({
        "token_id": token.token_id,
        "issuer": token.issuer,
        "subject": token.subject,
        "scope": token.scope,
        "purpose": token.purpose.model_dump(),
        "parent": token.parent,
        "expires": token.expires.isoformat(),
        "constraints": token.constraints.model_dump(),
    })


def get_token(token_id: str) -> DelegationToken | None:
    """Load a delegation token from the persistent store."""
    data = db_load_token(token_id)
    if data is None:
        return None
    return DelegationToken(**data)


def get_chain(token: DelegationToken) -> list[DelegationToken]:
    """Walk the DAG upward from a token to the root, returning the full chain."""
    chain = [token]
    current = token
    while current.parent is not None:
        parent = get_token(current.parent)
        if parent is None:
            break
        chain.append(parent)
        current = parent
    return list(reversed(chain))  # root first


def get_root_principal(token: DelegationToken) -> str:
    """Get the root principal (human) from a delegation chain."""
    chain = get_chain(token)
    return chain[0].issuer


def get_chain_token_ids(token: DelegationToken) -> list[str]:
    """Get the list of token IDs in the delegation chain (for audit logging)."""
    return [t.token_id for t in get_chain(token)]


def validate_delegation(
    token: DelegationToken,
    minimum_scope: list[str],
    capability_name: str,
) -> ANIPFailure | None:
    """Validate a delegation token for invoking a capability.

    Returns None if valid, or an ANIPFailure describing what's wrong.
    """
    # 1. Check expiry
    if token.expires < datetime.now(timezone.utc):
        return ANIPFailure(
            type="token_expired",
            detail=f"delegation token {token.token_id} expired at {token.expires.isoformat()}",
            resolution=Resolution(
                action="request_new_delegation",
                grantable_by=token.issuer,
            ),
            retry=True,
        )

    # 2. Check scope — the token must carry ALL required scopes (prefix match)
    token_scope_bases = [s.split(":")[0] for s in token.scope]  # "travel.book:max_$500" → "travel.book"
    missing_scopes = []
    for required_scope in minimum_scope:
        scope_matched = any(
            scope_base == required_scope or required_scope.startswith(scope_base + ".")
            for scope_base in token_scope_bases
        )
        if not scope_matched:
            missing_scopes.append(required_scope)
    if missing_scopes:
        root_principal = get_root_principal(token)
        return ANIPFailure(
            type="insufficient_authority",
            detail=f"delegation chain lacks scope(s): {', '.join(missing_scopes)}",
            resolution=Resolution(
                action="request_scope_grant",
                requires=f"delegation.scope += {', '.join(missing_scopes)}",
                grantable_by=root_principal,
            ),
            retry=True,
        )

    # 3. Check purpose binding — token purpose must match the capability
    if token.purpose.capability != capability_name:
        return ANIPFailure(
            type="purpose_mismatch",
            detail=(
                f"delegation token purpose is {token.purpose.capability} "
                f"but request is for {capability_name}"
            ),
            resolution=Resolution(
                action="request_new_delegation",
                grantable_by=token.issuer,
            ),
            retry=True,
        )

    # 4. Check delegation depth — walk the chain and count
    chain = get_chain(token)
    max_depth = token.constraints.max_delegation_depth
    # Depth is number of delegations (edges), not nodes
    actual_depth = len(chain) - 1
    if actual_depth > max_depth:
        return ANIPFailure(
            type="delegation_depth_exceeded",
            detail=f"delegation chain depth is {actual_depth}, max allowed is {max_depth}",
            resolution=Resolution(
                action="reduce_delegation_depth",
                requires=f"max_delegation_depth >= {actual_depth}",
                grantable_by=get_root_principal(token),
            ),
            retry=True,
        )

    # 5. Enforce concurrent_branches — reject if exclusive and another request is active
    if token.constraints.concurrent_branches == ConcurrentBranches.EXCLUSIVE:
        root = get_root_principal(token)
        if root in _active_requests:
            return ANIPFailure(
                type="concurrent_request_rejected",
                detail=f"concurrent_branches is exclusive and another request from {root} is in progress",
                resolution=Resolution(
                    action="wait_and_retry",
                    grantable_by=root,
                ),
                retry=True,
            )

    # 6. Validate parent chain — every parent must also be valid and not expired
    for ancestor in chain[:-1]:  # all except the current token
        if ancestor.expires < datetime.now(timezone.utc):
            return ANIPFailure(
                type="parent_token_expired",
                detail=f"ancestor token {ancestor.token_id} in delegation chain has expired",
                resolution=Resolution(
                    action="refresh_delegation_chain",
                    grantable_by=ancestor.issuer,
                ),
                retry=True,
            )

    return None  # all checks passed


def acquire_exclusive_lock(token: DelegationToken) -> None:
    """Mark a root principal as having an active request."""
    if token.constraints.concurrent_branches == ConcurrentBranches.EXCLUSIVE:
        _active_requests.add(get_root_principal(token))


def release_exclusive_lock(token: DelegationToken) -> None:
    """Release the active request lock for a root principal."""
    if token.constraints.concurrent_branches == ConcurrentBranches.EXCLUSIVE:
        _active_requests.discard(get_root_principal(token))


def validate_scope_narrowing(token: DelegationToken) -> ANIPFailure | None:
    """Validate that a child token's scope is a subset of its parent's scope.

    Returns None if valid (or no parent), or an ANIPFailure if scope widens.
    """
    if token.parent is None:
        return None  # root tokens have no parent to narrow from

    parent = get_token(token.parent)
    if parent is None:
        return None  # parent not registered — can't validate

    parent_scope_bases = {s.split(":")[0] for s in parent.scope}

    for child_scope in token.scope:
        child_base = child_scope.split(":")[0]
        # Child scope base must match or be narrower than a parent scope base
        matched = any(
            child_base == parent_base or child_base.startswith(parent_base + ".")
            for parent_base in parent_scope_bases
        )
        if not matched:
            return ANIPFailure(
                type="scope_escalation",
                detail=f"child token scope '{child_base}' is not a subset of parent token scopes: {', '.join(sorted(parent_scope_bases))}",
                resolution=Resolution(
                    action="narrow_scope",
                    requires=f"child scope must be subset of parent scope",
                    grantable_by=parent.issuer,
                ),
                retry=False,
            )

        # Check budget narrowing: child budget constraint must not exceed parent's
        if ":max_$" in child_scope:
            child_budget = float(child_scope.split(":max_$")[1])
            for parent_scope_str in parent.scope:
                parent_base = parent_scope_str.split(":")[0]
                if parent_base == child_base and ":max_$" in parent_scope_str:
                    parent_budget = float(parent_scope_str.split(":max_$")[1])
                    if child_budget > parent_budget:
                        return ANIPFailure(
                            type="scope_escalation",
                            detail=f"child budget ${child_budget} exceeds parent budget ${parent_budget} for scope '{child_base}'",
                            resolution=Resolution(
                                action="narrow_budget",
                                requires=f"budget must be <= ${parent_budget}",
                                grantable_by=parent.issuer,
                            ),
                            retry=False,
                        )

    return None


def check_budget_authority(token: DelegationToken, amount: float) -> ANIPFailure | None:
    """Check if the delegation chain carries sufficient budget authority."""
    for scope in token.scope:
        if ":max_$" in scope:
            max_budget = float(scope.split(":max_$")[1])
            if amount > max_budget:
                return ANIPFailure(
                    type="budget_exceeded",
                    detail=f"capability costs ${amount} but delegation chain authority is max ${max_budget}",
                    resolution=Resolution(
                        action="request_budget_increase",
                        requires=f"delegation.scope budget raised to ${amount}",
                        grantable_by=get_root_principal(token),
                    ),
                    retry=True,
                )
    return None

File: examples/anip-ts/src/primitives/delegation.ts
/**
 * Delegation chain validation.
 *
 * Validates the full delegation DAG: token expiry, scope sufficiency,
 * purpose binding, delegation depth, and parent chain.
 */

import { randomUUID } from "crypto";
import type { ANIPFailure, DelegationToken } from "../types.js";

// In-memory token store — maps token_id to DelegationToken
// In production this would be a database or token verification service
const tokenStore: Map<string, DelegationToken> = new Map();

// Active request tracking for concurrent_branches enforcement
const activeRequests: Set<string> = new Set();

export function validateParentExists(token: DelegationToken): ANIPFailure | null {
  /**
   * Validate that a token's parent exists in the store.
   * Returns null if valid (or no parent), or an ANIPFailure if the parent is missing.
   */
  if (token.parent === null) {
    return null; // root token
  }
  const parent = getToken(token.parent);
  if (parent === null) {
    return {
      type: "parent_not_found",
      detail: `parent token '${token.parent}' is not registered`,
      resolution: {
        action: "register_parent_token_first",
        requires: `token '${token.parent}' must be registered before its children`,
        grantable_by: token.issuer,
        estimated_availability: null,
      },
      retry: true,
    };
  }
  return null;
}

export function registerToken(token: DelegationToken): void {
  tokenStore.set(token.token_id, token);
}

export function getToken(tokenId: string): DelegationToken | null {
  return tokenStore.get(tokenId) ?? null;
}

export function getChain(token: DelegationToken): DelegationToken[] {
  /** Walk the DAG upward from a token to the root, returning the full chain. */
  const chain: DelegationToken[] = [token];
  let current = token;
  while (current.parent !== null) {
    const parent = getToken(current.parent);
    if (parent === null) {
      break;
    }
    chain.push(parent);
    current = parent;
  }
  return chain.reverse(); // root first
}

export function getRootPrincipal(token: DelegationToken): string {
  /** Get the root principal (human) from a delegation chain. */
  if (token.root_principal !== null) {
    return token.root_principal;
  }
  // Fallback for v0.1 tokens without root_principal field
  const chain = getChain(token);
  return chain[0].issuer;
}

export function resolveRegisteredToken(token: DelegationToken): DelegationToken | ANIPFailure {
  /**
   * Look up the stored version of a token and return it.
   * Returns the stored DelegationToken if the token_id is registered,
   * or an ANIPFailure if not. Callers MUST use the returned token for
   * all downstream operations — this prevents forged inline fields
   * (issuer, scope, constraints) from bypassing registration-time validation.
   */
  const stored = getToken(token.token_id);
  if (stored === null) {
    return {
      type: "token_not_registered",
      detail: `delegation token '${token.token_id}' is not registered — register via /anip/tokens first`,
      resolution: {
        action: "register_token",
        requires: "token must be registered before use",
        grantable_by: token.issuer,
        estimated_availability: null,
      },
      retry: true,
    };
  }
  return stored;
}

export function isANIPFailure(value: DelegationToken | ANIPFailure): value is ANIPFailure {
  return "type" in value && "detail" in value && "resolution" in value;
}

export function validateDelegation(
  token: DelegationToken,
  requiredScopes: string[],
  capabilityName: string
): DelegationToken | ANIPFailure {
  /**
   * Validate a delegation token for invoking a capability.
   * Returns the stored DelegationToken if valid (callers MUST use this for
   * all downstream operations), or an ANIPFailure describing what's wrong.
   */

  // 0. Resolve to stored token (prevents forged inline fields)
  const resolved = resolveRegisteredToken(token);
  if (isANIPFailure(resolved)) {
    return resolved;
  }
  token = resolved; // use stored token for all subsequent checks

  // 1. Check expiry
  const expiresDate = new Date(token.expires);
  if (expiresDate < new Date()) {
    return {
      type: "token_expired",
      detail: `delegation token ${token.token_id} expired at ${token.expires}`,
      resolution: {
        action: "request_new_delegation",
        grantable_by: getRootPrincipal(token),
        requires: null,
        estimated_availability: null,
      },
      retry: true,
    };
  }

  // 2. Check scope — the token must carry ALL required scopes (AND semantics)
  const missingScopes: string[] = [];
  for (const requiredScope of requiredScopes) {
    let scopeMatched = false;
    for (const scope of token.scope) {
      const scopeBase = scope.split(":")[0]; // "travel.book:max_$500" -> "travel.book"
      if (
        scopeBase === requiredScope ||
        requiredScope.startsWith(scopeBase + ".")
      ) {
        scopeMatched = true;
        break;
      }
    }
    if (!scopeMatched) {
      missingScopes.push(requiredScope);
    }
  }
  if (missingScopes.length > 0) {
    const rootPrincipal = getRootPrincipal(token);
    return {
      type: "insufficient_authority",
      detail: `delegation chain lacks scope: ${missingScopes.join(", ")}`,
      resolution: {
        action: "request_scope_grant",
        requires: `delegation.scope += ${missingScopes.join(", ")}`,
        grantable_by: rootPrincipal,
        estimated_availability: null,
      },
      retry: true,
    };
  }

  // 3. Check purpose binding — token purpose must match the capability
  if (token.purpose.capability !== capabilityName) {
    return {
      type: "purpose_mismatch",
      detail: `delegation token purpose is ${token.purpose.capability} but request is for ${capabilityName}`,
      resolution: {
        action: "request_new_delegation",
        grantable_by: getRootPrincipal(token),
        requires: null,
        estimated_availability: null,
      },
      retry: true,
    };
  }

  // 4. Verify the delegation chain is complete (no missing ancestors)
  const chain = getChain(token);
  // If the token has a parent but get_chain didn't reach a root (parent=null),
  // the chain is broken — an ancestor is unregistered
  if (chain[0].parent !== null) {
    return {
      type: "broken_delegation_chain",
      detail: `delegation chain is incomplete — ancestor token '${chain[0].parent}' is not registered`,
      resolution: {
        action: "register_missing_ancestor",
        grantable_by: getRootPrincipal(token),
        requires: null,
        estimated_availability: null,
      },
      retry: true,
    };
  }

  // 5. Check delegation depth
  const maxDepth = token.constraints.max_delegation_depth;
  // Depth is number of delegations (edges), not nodes
  const actualDepth = chain.length - 1;
  if (actualDepth > maxDepth) {
    return {
      type: "delegation_depth_exceeded",
      detail: `delegation chain depth is ${actualDepth}, max allowed is ${maxDepth}`,
      resolution: {
        action: "reduce_delegation_depth",
        requires: `max_delegation_depth >= ${actualDepth}`,
        grantable_by: getRootPrincipal(token),
        estimated_availability: null,
      },
      retry: true,
    };
  }

  // 6. Enforce concurrent_branches — reject if exclusive and another request is active
  if (token.constraints.concurrent_branches === "exclusive") {
    const root = getRootPrincipal(token);
    if (activeRequests.has(root)) {
      return {
        type: "concurrent_request_rejected",
        detail: `concurrent_branches is exclusive and another request from ${root} is in progress`,
        resolution: {
          action: "wait_and_retry",
          grantable_by: root,
          requires: null,
          estimated_availability: null,
        },
        retry: true,
      };
    }
  }

  // 7. Validate parent chain — every parent must also be valid and not expired
  for (const ancestor of chain.slice(0, -1)) {
    // all except the current token
    const ancestorExpires = new Date(ancestor.expires);
    if (ancestorExpires < new Date()) {
      return {
        type: "parent_token_expired",
        detail: `ancestor token ${ancestor.token_id} in delegation chain has expired`,
        resolution: {
          action: "refresh_delegation_chain",
          grantable_by: getRootPrincipal(token),
          requires: null,
          estimated_availability: null,
        },
        retry: true,
      };
    }
  }

  return token; // all checks passed — return stored token for downstream use
}

export function acquireExclusiveLock(token: DelegationToken): void {
  if (token.constraints.concurrent_branches === "exclusive") {
    activeRequests.add(getRootPrincipal(token));
  }
}

export function releaseExclusiveLock(token: DelegationToken): void {
  if (token.constraints.concurrent_branches === "exclusive") {
    activeRequests.delete(getRootPrincipal(token));
  }
}

export function validateScopeNarrowing(token: DelegationToken): ANIPFailure | null {
  /**
   * Validate that a child token's scope is a subset of its parent's scope.
   * Returns null if valid (or no parent), or an ANIPFailure if scope widens.
   */
  if (token.parent === null) {
    return null; // root tokens have no parent to narrow from
  }

  const parent = getToken(token.parent);
  if (parent === null) {
    return {
      type: "parent_not_found",
      detail: `parent token '${token.parent}' is not registered — cannot validate scope narrowing`,
      resolution: {
        action: "register_parent_token_first",
        requires: `token '${token.parent}' must be registered before its children`,
        grantable_by: token.issuer,
        estimated_availability: null,
      },
      retry: true,
    };
  }

  const parentScopeBases = new Set(parent.scope.map((s) => s.split(":")[0]));

  for (const childScope of token.scope) {
    const childBase = childScope.split(":")[0];

    // Child scope base must match or be narrower than a parent scope base
    let matched = false;
    for (const parentBase of parentScopeBases) {
      if (childBase === parentBase || childBase.startsWith(parentBase + ".")) {
        matched = true;
        break;
      }
    }

    if (!matched) {
      return {
        type: "scope_escalation",
        detail: `child token scope '${childBase}' is not a subset of parent token scopes: ${[...parentScopeBases].sort().join(", ")}`,
        resolution: {
          action: "narrow_scope",
          requires: "child scope must be subset of parent scope",
          grantable_by: getRootPrincipal(parent),
          estimated_availability: null,
        },
        retry: false,
      };
    }

    // Check budget constraints: if parent has a budget on this scope,
    // child MUST preserve it (same or tighter). Dropping it is escalation.
    for (const parentScopeStr of parent.scope) {
      const pBase = parentScopeStr.split(":")[0];
      if (pBase === childBase && parentScopeStr.includes(":max_$")) {
        const parentBudget = parseFloat(parentScopeStr.split(":max_$")[1]);
        if (!childScope.includes(":max_$")) {
          // Child dropped the budget constraint entirely
          return {
            type: "scope_escalation",
            detail: `child dropped budget constraint from scope '${childBase}' (parent has max $${parentBudget})`,
            resolution: {
              action: "preserve_budget_constraint",
              requires: `scope '${childBase}' must include budget <= $${parentBudget}`,
              grantable_by: getRootPrincipal(parent),
              estimated_availability: null,
            },
            retry: false,
          };
        }
        const childBudget = parseFloat(childScope.split(":max_$")[1]);
        if (childBudget > parentBudget) {
          return {
            type: "scope_escalation",
            detail: `child budget $${childBudget} exceeds parent budget $${parentBudget} for scope '${childBase}'`,
            resolution: {
              action: "narrow_budget",
              requires: `budget must be <= $${parentBudget}`,
              grantable_by: getRootPrincipal(parent),
              estimated_availability: null,
            },
            retry: false,
          };
        }
      }
    }
  }

  return null;
}

export function validateConstraintsNarrowing(token: DelegationToken): ANIPFailure | null {
  /**
   * Validate that a child token's constraints don't weaken its parent's.
   * Returns null if valid (or no parent), or an ANIPFailure if constraints widen.
   */
  if (token.parent === null) {
    return null;
  }

  const parent = getToken(token.parent);
  if (parent === null) {
    return null; // parent existence is checked separately
  }

  // max_delegation_depth: child cannot raise it
  if (token.constraints.max_delegation_depth > parent.constraints.max_delegation_depth) {
    return {
      type: "constraint_escalation",
      detail: `child max_delegation_depth (${token.constraints.max_delegation_depth}) exceeds parent (${parent.constraints.max_delegation_depth})`,
      resolution: {
        action: "narrow_constraints",
        requires: `max_delegation_depth must be <= ${parent.constraints.max_delegation_depth}`,
        grantable_by: getRootPrincipal(parent),
        estimated_availability: null,
      },
      retry: false,
    };
  }

  // concurrent_branches: child cannot weaken from exclusive to allowed
  if (
    parent.constraints.concurrent_branches === "exclusive" &&
    token.constraints.concurrent_branches === "allowed"
  ) {
    return {
      type: "constraint_escalation",
      detail: "child weakened concurrent_branches from 'exclusive' to 'allowed'",
      resolution: {
        action: "preserve_constraint",
        requires: "concurrent_branches must remain 'exclusive'",
        grantable_by: getRootPrincipal(parent),
        estimated_availability: null,
      },
      retry: false,
    };
  }

  return null;
}

export function issueToken(
  subject: string,
  scope: string[],
  capability: string,
  issuerId: string,
  parentToken: DelegationToken | null,
  purposeParameters: Record<string, unknown>,
  ttlHours: number,
  rootPrincipal: string | null = null,
): { token: DelegationToken; tokenId: string } {
  const tokenId = `anip-${randomUUID().replace(/-/g, "").slice(0, 12)}`;
  const now = new Date();
  const expires = new Date(now.getTime() + ttlHours * 3600 * 1000);

  let maxDepth = 3;
  let concurrent: "allowed" | "exclusive" = "allowed";
  if (parentToken !== null) {
    maxDepth = Math.min(maxDepth, parentToken.constraints.max_delegation_depth);
    concurrent = parentToken.constraints.concurrent_branches;
  }

  const token: DelegationToken = {
    token_id: tokenId,
    issuer: issuerId,
    subject,
    scope,
    purpose: {
      capability,
      parameters: purposeParameters,
      task_id: `task-${tokenId}`,
    },
    parent: parentToken?.token_id ?? null,
    expires: expires.toISOString(),
    constraints: {
      max_delegation_depth: maxDepth,
      concurrent_branches: concurrent,
    },
    root_principal: rootPrincipal,
  };

  // Validate narrowing if child token
  if (parentToken !== null) {
    const scopeFailure = validateScopeNarrowing(token);
    if (scopeFailure !== null) {
      throw new Error(scopeFailure.detail);
    }
    const constraintFailure = validateConstraintsNarrowing(token);
    if (constraintFailure !== null) {
      throw new Error(constraintFailure.detail);
    }
  }

  registerToken(token);
  return { token, tokenId };
}

export function getChainTokenIds(token: DelegationToken): string[] {
  return getChain(token).map((t) => t.token_id);
}

export function checkBudgetAuthority(
  token: DelegationToken,
  amount: number
): ANIPFailure | null {
  /** Check if the delegation chain carries sufficient budget authority. */
  for (const scope of token.scope) {
    if (scope.includes(":max_$")) {
      const maxBudget = parseFloat(scope.split(":max_$")[1]);
      if (amount > maxBudget) {
        return {
          type: "budget_exceeded",
          detail: `capability costs $${amount} but delegation chain authority is max $${maxBudget}`,
          resolution: {
            action: "request_budget_increase",
            requires: `delegation.scope budget raised to $${amount}`,
            grantable_by: getRootPrincipal(token),
            estimated_availability: null,
          },
          retry: true,
        };
      }
    }
  }
  return null;
}
